server/src/main/java/org/candlepin/model/OwnerContentCurator.java

```diff
@@ -334,96 +334,22 @@ public int clearContentForOwner(Owner owner) {
-     * <em>except</em> the org specified. If no content is found for the specified IDs in other
-     * orgs, this method returns an empty map.
-     *
-     * @param owner
-     *  The owner to exclude from the content lookup. If this value is null, no owner-filtering
-     *  will be performed.
-     *
-     * @param contentIds
-     *  A collection of content IDs for which to fetch all known versions
-     *
-     * @return
-     *  A map containing all known versions of the given content, mapped by Red Hat ID
-     */
-    public Map<String, Set<Content>> getVersionedContentById(Owner owner, Collection<String> contentIds) {
-        Map<String, Set<Content>> result = new HashMap<>();
-
-        if (contentIds != null && !contentIds.isEmpty()) {
-            String jpql;
-
-            if (owner != null) {
-                jpql = "SELECT c FROM OwnerContent oc JOIN oc.content c " +
-                    "WHERE oc.owner.id != :owner_id AND c.id IN (:cids)";
-            }
-            else {
-                jpql = "SELECT c FROM Content c WHERE c.id IN (:cids)";
-            }
-
-            TypedQuery<Content> query = this.getEntityManager().createQuery(jpql, Content.class);
-
-            if (owner != null) {
-                query.setParameter("owner_id", owner.getId());
-            }
-
-            for (Collection<String> block : this.partition(contentIds)) {
-                List<Content> fetched = query.setParameter("cids", block)
-                    .getResultList();
-
-                for (Content entity : fetched) {
-                    Set<Content> idSet = result.get(entity.getId());
-
-                    if (idSet == null) {
-                        idSet = new HashSet<>();
-                        result.put(entity.getId(), idSet);
-                    }
-
-                    idSet.add(entity);
-                }
-            }
-        }
-
-        return result;
-    }
-
-    /**
-     * Fetches all content belonging to organizations other than the one provided having an entity
-     * version equal to one of the versions provided.
-     *
-     * @param exclude
-     *  The owner/organization to exclude from the search. If null, no organizations will be
-     *  excluded
+     * Fetches all content having an entity version equal to one of the versions provided.
      *
      * @param versions
      *  A collection of entity versions to use to select contents
      *
      * @return
      *  a map containing the contents found, keyed by content ID
      */
-    public Map<String, List<Content>> getContentByVersions(Owner exclude, Collection<Integer> versions) {
+    public Map<String, List<Content>> getContentByVersions(Collection<Integer> versions) {
         Map<String, List<Content>> result = new HashMap<>();
 
         if (versions != null && !versions.isEmpty()) {
-            TypedQuery<Content> query;
+            String jpql = "SELECT c FROM Content c WHERE c.entityVersion IN (:vblock)";
```

Comment:
One more thing just came to mind: Now that we can reuse orphaned entities, what happens if the OrphanCleanupJob runs at the same time? I see the job is using pessimistic write lock, so if it manages to lock the entity first this side might fail to read with a locktimeoutexception, but what if the job acquires the lock after this query is done, but right before we create the new owner-content mapping? Should we be locking entities in this side too?

Reply:
Transaction isolation and referential integrity will save us here, so long as we continue to only create owner-entity mappings in transactions and don't lift that foreign key ref. 

In the case of parallel ops here, we have the following cases:

1. Orphan cleanup gets the lock first. When this happens, the product mapping portion of refresh has either already passed, or is waiting to run. A potential lockwait failure can occur, but by the time the orphan cleanup job is complete and the lock is released, the product will no longer exist and refresh will not see it.
2. Refresh gets the lock first. When this occurs, refresh is either creating a new reference to the product, or removing it from the owner. In the former case, once the lock is acquired in orphan cleanup, the query shouldn't return it. For the later case, it won't matter.

This is largely off the top of my head, so some particulars of how locks propagate with foreign keys, and whether or not the lock acquired by the orphan cleanup job query will acquire it before or after determining the results of the query are a bit fuzzy. Even in the worst-case, I expect that a foreign key violation will save us from doing anything bad.